servers/src/grin/seed.rs

```diff
@@ -256,18 +271,58 @@ fn listen_for_addrs(
 			let _ = thread::Builder::new()
 				.name("peer_connect".to_string())
 				.spawn(move || {
-					let connect_peer = p2p_c.connect(&addr);
-					match connect_peer {
-						Ok(p) => {
-							trace!(LOGGER, "connect_and_req: ok. attempting send_peer_request");
-							if let Ok(p) = p.try_read() {
-								let _ = p.send_peer_request(capab);
+					let mut connect_retry_count = 0;
+					loop {
+						let connect_peer = p2p_c.connect(&addr);
+						match connect_peer {
+							Ok(p) => {
+								trace!(
+									LOGGER,
+									"connect_and_req: from {}:{} to {}, ok. attempting send_peer_request",
+									p2p_c.config.host,
+									p2p_c.config.port,
+									addr
+								);
+								if let Ok(p) = p.try_read() {
+									let _ = p.send_peer_request(capab);
+								}
+								let _ = peers_c.update_state(addr, p2p::State::Healthy);
+								break;
+							}
+							Err(e) => {
+								debug!(
+									LOGGER,
+									"connect_and_req: from {}:{} to {}, is Defunct. {:?}",
+									p2p_c.config.host,
+									p2p_c.config.port,
+									addr,
+									e,
+								);
+								let _ = peers_c.update_state(addr, p2p::State::Defunct);
+
+								// don't retry if connection refused
+								if let p2p::Error::Connection(io_err) = e {
+									if io::ErrorKind::ConnectionRefused == io_err.kind() {
+										break;
+									}
+								}
 							}
 						}
-						Err(e) => {
-							debug!(LOGGER, "connect_and_req: {} is Defunct; {:?}", addr, e);
-							let _ = peers_c.update_state(addr, p2p::State::Defunct);
+
+						// retry for 3 times
+						thread::sleep(time::Duration::from_secs(1));
```

Comment:
Perhaps sleep for `connect_retry_count` or `2^ connect_retry_count`?

Reply:
Perhaps no good to sleep too long, 1 seconds and retry 3 times should be enough to skip the unexpected exception but the peer is indeed in well condition.

In practical Grin network, one peer connection fail is not a major issue, we just skip to another peer.  This retry could be only meaningful for `travis-ci` test environment.